### PR TITLE
3897 commit message template file not found

### LINKED ISSUE #3897: commit message template file not found in root folder of repo
**Do you want to request a *feature* or report a *bug*?**
A bug

**What is the current behavior?**
I use a commit message template file.
(idea from http://codeinthehole.com/tips/a-useful-template-for-commit-messages/)

Git Extension shows an Exception dialog when I try to commit.
screenshot-000022

**If the current behavior is a bug, please provide the steps to reproduce and if possible a minimal demo of the problem.**

I have this line in my .git\config file:
```
[commit]
	template = .gitmessage.txt
```
There is a .gitmessage.txt file in the root of my repository.

**What is the expected behavior?**
When I commit from the cmd line, the file is found at the root of my repository, and listed as a starting point for the commit message in the editor that opens.

Git Extensions should not look in 'C:\Program Files (x86)\GitExtensions\'.
But in the root folder of the repository

**Environment you encounter the issue:**

- GitExtensions version: GitExtensions.exe [2.50.01]
- GIT version: git version 2.13.0.windows.1
- OS version: Win 10 Enterprise

**Did this work in previous version of GitExtensions (which)?**
No idea.

### LINKED ISSUE #3987: F-Secure virus alert from 2.50.02 MSI
For your information that F-Secure Client Security Premium reports that GitExtensions-2.50.02-SetupComplete.msi installer contains a _Gen:Variant.Razy.158435_ in GitExtSshAskPass.exe executable.

Here's a VirusTotal report: https://www.virustotal.com/#/file/1321116c5998a4ce6b87a6c3e7dd7396bd1109d55ef26a3e62d73dd5a3484dce/detection

### PR DESCRIPTION
**Pull request by RussKie**
Fixes #3897.

Changes proposed in this pull request:
 - Move `/GitUI/CommandsDialogs/CommitDialog/CommitTemplateItem.cs` to `/GitCommands/CommitTemplateItem.cs`
- Create `/GitCommands/CommitTemplateManager.cs` to load commit message template, if one is specified
- `CommitTemplateManager` can be later used to migrate serialization implementations from `CommitTemplateItem` (`CommitTemplateItem` is a dto, shouldn't have any behaviors attached to it).

The current state:

screen shot 2017-08-19 at 1 35 04 pm

The new state:
screen shot 2017-08-19 at 1 36 11 pm

### CODE REVIEW

**Comment by RussKie** [comment]
In the light on discussion in #3898 I would like to know the general feeling about this.
I haven't written any tests yet, holding off until get conceptual agreement on the approach.

@jbialobr @sharwell @pmiossec 

I am also a bit unsure about this - https://github.com/gitextensions/gitextensions/blob/2d15fc1a68f999d56f08326e759becb35fdbb1f9/GitUI/CommandsDialogs/CommitDialog/CommitTemplateItem.cs#L119
The change was introduced about 4 years ago https://github.com/gitextensions/gitextensions/commit/8a37839ebdc6c9dc2a127fb39b99316669ce0836#diff-3fe0b06af977c3d151c2d7798e280450R115, unfortunately there was no tests added so it is hard to ensure there is no regressions...

**Comment by jbialobr** [comment]
>I am also a bit unsure about this -

It looks like it is used to migrate from the old data format. I think that after 4 years we can remove the migration logic.

File: GitCommands/CommitTemplateManager.cs [hunk 1]
```diff
@@ -0,0 +1,85 @@
...
+        /// Loads commit template from the file specified in .git/config 
+        /// under <c>commit.template</c> setting. 
+        /// </summary>
+        /// <exception cref="FileNotFoundException">The specified template file cannot be found.</exception>
+        /// <returns>The commit template, if it is specified; otherwise <see langword="null"/>.</returns>
+        /// <remarks>
+        /// Template file can be set by the following command:
+        /// <c>$ git config --global commit.template ~/.git_commit_msg.txt</c>
+        /// </remarks>
+        public string LoadGitCommitTemplate(string workingDir)
```
Comment (jbialobr) [inline GitCommands/CommitTemplateManager.cs]: Do we need to pass `workingDir` having `_module`?

File: GitCommands/CommitTemplateManager.cs [hunk 2]
```diff
@@ -0,0 +1,85 @@
...
+
+            if (!Path.IsPathRooted(fileName))
+            {
+                // git automatically converts '~' if the path contains it
+                // i.e. ~/.template.txt -> C:/Users/<user>/.template.txt (Windows)
+                //                      -> /home/<user>/.gitcommit.txt (Linux)
+                // however if the .gitconfig contains a template starting with '~'
+                // it means the user has manually entered the path
+                // in this case, treat the path as local
+                if (fileName.StartsWith("~"))
```
Comment (jbialobr) [inline GitCommands/CommitTemplateManager.cs]: Why would we like to change the standard git behavior?

File: GitUI/CommandsDialogs/FormCommit.cs [hunk 3]
```diff
@@ -1191,79 +1210,79 @@ private void Unstage(bool canUseUnstageAll = true)
...
+                toolStripProgressBar1.Value = 0;
+                Staged.StoreNextIndexToSelect();
 
-                    var files = new List<GitItemStatus>();
-                    var allFiles = new List<GitItemStatus>();
+                var files = new List<GitItemStatus>();
+                var allFiles = new List<GitItemStatus>();
 
-                    foreach (var item in Staged.SelectedItems)
+                foreach (var item in Staged.SelectedItems)
```
Comment (jbialobr) [inline GitUI/CommandsDialogs/FormCommit.cs]: It is nice to refactor in a separate PR.

File: GitCommands/CommitTemplateManager.cs (hunk 2 above)
Comment (RussKie) [inline GitCommands/CommitTemplateManager.cs]: this is to handle non standard behavior
if you add the template via git bash, the `~` gets translated into the physical path
this implementation is to handle when user decides to modify the config manually and entering unresolved path

File: GitUI/CommandsDialogs/FormCommit.cs (hunk 3 above)
Comment (RussKie) [inline GitUI/CommandsDialogs/FormCommit.cs]: this is VS reformatting the whitespaces...
if you add `?w=1` to the url you won't see them

File: GitCommands/CommitTemplateManager.cs (hunk 1 above)
Comment (RussKie) [inline GitCommands/CommitTemplateManager.cs]: good point, I've missed it somehow

File: GitCommands/CommitTemplateManager.cs (hunk 2 above)
Comment (jbialobr) [inline GitCommands/CommitTemplateManager.cs]: It is almost sure, that there is no such file in the repository. The user will get a FNFE telling him that the file does not exist and giving the full path <repo path>/<fileName>. This will be quite surprising. If you don't want to expand the user home dir (don't know why), then show the original path from the config file.

Comment (RussKie) [inline GitCommands/CommitTemplateManager.cs]: For those doing a lot of web dev it may not be _that_ surprising... In asp.net `~` has specific meaning - [the root of the app](http://www.cryer.co.uk/brian/csharp/tilde_virtual_paths_in_asp_net.htm).

But I'm happy to change it to throw FNFE

Comment (jbialobr) [inline GitCommands/CommitTemplateManager.cs]: I find it far fetched and would rather stick to the git/file system semantic.

https://git-scm.com/book/tr/v2/Customizing-Git-Git-Configuration

```
To tell Git to use it as the default message that appears in your editor 
when you run git commit, set the commit.template configuration value:

$ git config --global commit.template ~/.gitmessage.txt
$ git commit
```

File: GitUI/CommandsDialogs/FormCommit.cs (hunk 3 above)
Comment (jbialobr) [inline GitUI/CommandsDialogs/FormCommit.cs]: But then I am not able to leave a comment.

Comment (RussKie) [inline GitUI/CommandsDialogs/FormCommit.cs]: Sneaky github...

File: GitCommands/CommitTemplateManager.cs [hunk 4]
```diff
@@ -0,0 +1,75 @@
...
+        /// </remarks>
+        public string LoadGitCommitTemplate()
+        {
+            string fileName = _module.GetEffectiveSetting("commit.template");
+            if (string.IsNullOrEmpty(fileName))
+            {
+                return null;
+            }
+
+            if (!Path.IsPathRooted(fileName))
```
Comment (jbialobr) [inline GitCommands/CommitTemplateManager.cs]: This expanding to the full path logic looks like out of place here. It should be externalized to be reusable. Perhaps there could be handled the '~' case as a general rule.

File: GitCommands/CommitTemplateManager.cs [hunk 5]
```diff
@@ -0,0 +1,75 @@
...
+            {
+                var path = Path.GetFullPath(Path.Combine(_module.WorkingDir ?? "", fileName));
+                var uri = new Uri(path);
+                fileName = uri.LocalPath;
+            }
+
+            if (!_fileSystem.File.Exists(fileName))
```
Comment (jbialobr) [inline GitCommands/CommitTemplateManager.cs]: We have a function for this. It handles cases when the given path contains invalid chars.
https://github.com/gitextensions/gitextensions/blob/e81a6c1c94657361c11cf6b727807f2653d29023/GitCommands/PathUtil.cs#L172

File: UnitTests/GitCommandsTests/CommitTemplateManagerTests.cs [hunk 6]
```diff
@@ -0,0 +1,68 @@
...
+        {
+            _module.GetEffectiveSetting("commit.template").Returns(template);
+
+            _manager.LoadGitCommitTemplate().Should().BeNull();
+        }
+
+        [TestCase("~/template.txt")]
+        [TestCase("./template.txt")]
+        [TestCase("template.txt")]
+        public void LoadGitCommitTemplate_path_not_rooted(string template)
```
Comment (jbialobr) [inline UnitTests/GitCommandsTests/CommitTemplateManagerTests.cs]: LoadGitCommitTemplate_Throws_WhenTemplateFileNotExists

File: UnitTests/GitCommandsTests/CommitTemplateManagerTests.cs [hunk 7]
```diff
@@ -0,0 +1,68 @@
...
+        {
+            const string workingDir = @"c:\dev\repo";
+            _module.WorkingDir.Returns(workingDir);
+            _module.GetEffectiveSetting("commit.template").Returns(template);
+
+            ((Action)(() => _manager.LoadGitCommitTemplate())).ShouldThrow<FileNotFoundException>();
+        }
+
+        [Test]
+        public void LoadGitCommitTemplate_should_load_file_content()
```
Comment (jbialobr) [inline UnitTests/GitCommandsTests/CommitTemplateManagerTests.cs]: I see here 3 tests combined:
1. LoadGitCommitTemplate_ExpandsRelativePathToWorkingDirSubpath
2. LoadGitCommitTemplate_should_load_file_content
3. LoadGitCommitTemplate_RmovesCRFormFileContent

File: UnitTests/GitCommandsTests/CommitTemplateManagerTests.cs [hunk 8]
```diff
@@ -0,0 +1,68 @@
...
+        {
+            const string workingDir = @"c:\dev\repo";
+            _module.WorkingDir.Returns(workingDir);
+            _module.GetEffectiveSetting("commit.template").Returns("./template.txt");
+            _file.Exists(Arg.Any<string>()).Returns(true);
+            _file.ReadAllText(Arg.Any<string>()).Returns("line1\r\nline2\rline3\nline4");
```
Comment (jbialobr) [inline UnitTests/GitCommandsTests/CommitTemplateManagerTests.cs]: If I change the controller implementation to:
```
string commitTemplate = _fileSystem.File.ReadAllText(fileName, _module.CommitEncoding).Replace("\r", "");
```
then the test will fail.

Comment (RussKie) [reply UnitTests/GitCommandsTests/CommitTemplateManagerTests.cs]: and it must - you've just changed the behavior.

Comment (jbialobr) [reply UnitTests/GitCommandsTests/CommitTemplateManagerTests.cs]: I think that we are not able to find a compromise here. When tests are written to test the implementation, then yes, it must. But when they are written to verify assumptions (pre and post conditions like given-when-then) then it should not.

Comment (jbialobr) [reply UnitTests/GitCommandsTests/CommitTemplateManagerTests.cs]: If in some point in time we decided to replace
`string fileName = _module.GetEffectiveSetting("commit.template");`
with 
```
if (!_module.EffectiveConfigFile.HasValue("commit.template"))
  return;
string fileName = _module.EffectiveConfigFile.GetValue("commit.template");
```
then again the test will fail.

Comment (RussKie) [reply UnitTests/GitCommandsTests/CommitTemplateManagerTests.cs]: yes, that is correct - you have changed the contract

Comment (jbialobr) [reply UnitTests/GitCommandsTests/CommitTemplateManagerTests.cs]: How is that? The LoadGitCommitTemplate_commit_template_not_defined still passes. The two other tests don't.

Comment (RussKie) [reply UnitTests/GitCommandsTests/CommitTemplateManagerTests.cs]: > But when they are written to verify assumptions (pre and post conditions like given-when-then) then it should not.

You have to arrange before the assertion. The arrangement is based on an existing contract.

How would you test it?

File: UnitTests/GitCommandsTests/CommitTemplateManagerTests.cs (hunk 7 above)
Comment (RussKie) [reply UnitTests/GitCommandsTests/CommitTemplateManagerTests.cs]: Yes, if you insist on nitpicking :)

File: UnitTests/GitCommandsTests/CommitTemplateManagerTests.cs (hunk 8 above)
Comment (jbialobr) [reply UnitTests/GitCommandsTests/CommitTemplateManagerTests.cs]: Create a config file (in memory or on hard disk) that contains `commit.template` entry.

Comment (RussKie) [reply UnitTests/GitCommandsTests/CommitTemplateManagerTests.cs]: The way I see it - any tests are better than no tests at all. 
Currently the test coverage is very limited and academic discussions how a very unlikely change in a possible future may break a test is counter productive.

If you modify a contract any functionality dependent on that contract will be affected and there is a chance of regression. Of course you will need to test it. 
I would be very worried if tests wouldn't fail in such case.

File: GitCommands/CommitTemplateManager.cs (hunk 4 above)
Comment (RussKie) [reply GitCommands/CommitTemplateManager.cs]: I'm not sure I can follow your idea.
Do you think you could write some (pseudo)code to help visualising what you mean?

Comment (RussKie) [reply GitCommands/CommitTemplateManager.cs]: I'm not sure I can follow your idea.
Do you think you could write some (pseudo)code to help visualising what you mean?

File: GitCommands/CommitTemplateManager.cs (hunk 5 above)
Comment (RussKie) [reply GitCommands/CommitTemplateManager.cs]: the existing method won't work because it requires physical access to filesystem
we'll need to abstract all filesystem routines first before we can reuse them

File: UnitTests/GitCommandsTests/CommitTemplateManagerTests.cs (hunk 7 above)
Comment (jbialobr) [reply UnitTests/GitCommandsTests/CommitTemplateManagerTests.cs]: No, I am only a human, so do other devs and we all make mistakes. When there are no explicit assumptions given, then it is hard to reason if the test author wrote the test correctly (if the test implementation is correct according to the spec).

File: UnitTests/GitCommandsTests/CommitTemplateManagerTests.cs (hunk 8 above)
Comment (RussKie) [reply UnitTests/GitCommandsTests/CommitTemplateManagerTests.cs]: > Create a config file (in memory or on hard disk) that contains commit.template entry.

How would it change things?

Also this will not be just testing the `CommitTemplateManager` but also `GitModule` reading settings, then `FileSystem` checking and loading files.
This is wrong

Comment (jbialobr) [reply UnitTests/GitCommandsTests/CommitTemplateManagerTests.cs]: Perhaps we have to define the meaning of `contract` more precisely as I don't see any change in the contract here.

>The way I see it - any tests are better than no tests at all.

Sure. But the discussion is at the general level and it should happen in https://github.com/gitextensions/gitextensions/pull/3898. The intention is to increase the test coverage and the discussion is on how to do it properly (as we are at the beginning of that process). 
I know that the change I gave as an example is unlikely to happen. It was only to illustrate how fragile tests written in this manner are.

Comment (jbialobr) [reply UnitTests/GitCommandsTests/CommitTemplateManagerTests.cs]: >How would it change things?

I am not able to explain it more. Go implement and you will see.

>This is wrong

As you see it is disputable. Such an approach has an added value. You cross-verify that the assumptions you made on implementation and external interfaces are good. Read https://medium.com/python-pandemonium/mock-or-not-to-mock-41965d33f175 as an example.

Comment (RussKie) [reply UnitTests/GitCommandsTests/CommitTemplateManagerTests.cs]: An interesting read, but I didn't find it convincing

Comment (jbialobr) [reply UnitTests/GitCommandsTests/CommitTemplateManagerTests.cs]: I feel like I am wasting my time. I can not reason against `I didn't find it convincing`
You didn't answer to `How is that?` but then you continue conversation like it was agreed that I changed the contract: `If you modify a contract any functionality dependent on that contract will be affected and there is a chance of regression.`

File: GitUI/CommandsDialogs/FormCommit.cs (hunk 3 above)
Comment (RussKie) [reply GitUI/CommandsDialogs/FormCommit.cs]: extracted as a separate commit

**Comment by RussKie** [comment]
Ok we aren't getting anywhere and the level of frustration is rising...

Let's start afresh.
I've changed functionality and I want to ensure it is tested to the point I
feel confident using it everywhere else knowing it is correct. I also want
to be sure that any changes to the implementation within the method have
little or no effect on consumers of this method.

I've provided a test coverage in the way I know. It is certainly not the
only way, and it may not the best. However it works.

How would you test the functionality to provide the same level of
confidence?

I would appreciate a real example so we can discuss pros and cons of both
mine and your implementations.

**Comment by jbialobr** [comment]
Here you are https://github.com/RussKie/gitextensions/pull/4

Of course the TestUtl is only a proof of concept and requires more work. It is based on https://github.com/libgit2/libgit2sharp

Surprisingly some tests run faster when not mocked (those prefixed with X). 
gitext

**Comment by jbialobr** [comment]
I do a refactoring (lets put aside if it makes any sense to do it): https://github.com/gitextensions/gitextensions/pull/3940
The change is semantically insignificant. I run my tests to verify that everything is ok, but it turns out it is not. 6 out of 8 tests fail. Why I would have to correct them? How can I know that I did not introduce any regression having to change 6 out of 8 tests? Such tests does not play their fundamental role.

**Comment by jbialobr** [comment]
https://github.com/gitextensions/gitextensions/pull/3941 - All I would like to do is to add a new test that ensures the proper encoding is used. The other tests should not fail.

https://github.com/gitextensions/gitextensions/pull/3942 - I would expect that tests will catch such a mistake. This is the lack of completeness I was talking about in https://github.com/gitextensions/gitextensions/pull/3898

**Comment by jbialobr** [comment]
>the existing method won't work because it requires physical access to filesystem
we'll need to abstract all filesystem routines first before we can reuse them

https://github.com/gitextensions/gitextensions/commit/6ed75cbcbc942970d9b72d2bb1a4cbf1a669ca22 - mocked 5/8, not mocked 8/8

**Comment by jbialobr** [comment]
>I'm not sure I can follow your idea.
Do you think you could write some (pseudo)code to help visualising what you mean?

https://github.com/gitextensions/gitextensions/commit/bd4f8e6be73786c493aaa75c16c566f5515c060e

**Comment by RussKie** [comment]
Thank you for the effort in demonstrating possible issues. 

0f98e0241b74809db49e2d0f708bd72c536cd289, 7ed9553ddd0c5ce4167cbcc984ba4747d7f41d7c - the failed tests are expected, the implementation has changed. No suprises here.

8f5cb1aba2d0a8d3298d83a446d02252c399aa49 - the change is superficial, but illustrates your point quite well. 👍 
But I don't think any test is 100% fool proof...

6ed75cbcbc942970d9b72d2bb1a4cbf1a669ca22 - whilst the refactor looks simple, it is technically incorrect - the manager's implementation now depends on a concrete `PathService` implementation and the test turns into an integration test. A correct implementation would inject `IPathService` instead of `IFilsSystem` then have the required methods mocked.


I've looked at 9fa98b89872c89dffdefb8c56eafa47846dd5c79, thank you
You version works too 😄 This probably matters the most.
Though, I have to admit, I find it much harder to write tests your way.

**Comment by jbialobr** [comment]
>0f98e02, 7ed9553 - the failed tests are expected, the implementation has changed. No suprises here.

The surprise is here. Especially for 0f98e02. I second my questions: `Why I would have to correct them? How can I know that I did not introduce any regression having to change 6 out of 8 tests?`

>But I don't think any test is 100% fool proof...

Sure, but we want to choose more robust approach. I wouldn't like to make any assumptions about which mistakes we are allowed to make. 

>6ed75cb - whilst the refactor looks simple, it is technically incorrect

What's new? I did it this way only to not have to refactor your tests. You asked me to  `write some (pseudo)code to help visualising what I mean`. Notice that my tests go green and do need no change even if you do the refactoring the correct way.

>This probably matters the most.

My version works where yours does not. I don't observe the opposite situation.
My version allows us to write tests for the existing code base. Then we can refactor it expecting all tests go green. I don't think we should do it the other way around. I consider automated tests as a guard that ensures the expected behavior does not change while the code base changes (for various reasons).

I am looking forward to the pros and cons discussion you promised.

**Comment by RussKie** [comment]
I will try to explain how I approach design and implementation tasks. This is reflective how we do development and engage with offshore partners.

To start with, let's reiterate the issue.

_A user must be able to specify a commit template._
1. _If a commit template is not in an absolute URI, it should be assumed as a relative to the current working directory._
2. _If a commit template does not exist the app must not crash, and show an error message explaining the error._

Looking at the current implementation it tries doing all of that and fails along the way.

There are two distinct requirements here.
The second one is best placed in FormCommit, since it represents the UI layer. 
To address the requirement (keeping in mind OO design principles) the existing API is getting changed to something like the following:

```
   try
   {
       Message.Text = _commitTemplateManager.LoadGitCommitTemplate();
       return;
   }
   catch (FileNotFoundException ex)
   {
       MessageBox.Show("error");
   }
   catch (Exception ex)
   {
       MessageBox.Show("error");
   }
   Message.Text = string.Empty;
  
```

From the design perspective, as an API consumer by calling `ICommitTemplateManager.LoadGitCommitTemplate()` I either 
* get the template content, or
* get FileNotFoundException.
Add general exception handler just for completeness.
This aligns well with [SOLID](https://en.wikipedia.org/wiki/SOLID_(object-oriented_design)) (open/close, interface segregation, dependency inversion) and [GRASP](https://en.wikipedia.org/wiki/GRASP_(object-oriented_design)) (information expert, low coupling, high cohesion).

The first requirement is to be addressed by `LoadGitCommitTemplate()` implementation. 
To recap, it must address the following requirements:
* _if template is not specified - return null_
* _if template is specified and it is not an absolute URI - treat it as relative to the current working directory_
* _if template is specified and file does not exist - throw FileNotFoundException providing relevant information_
* _else return the template content._


This is a pretty trivial example, but in production environment a lot of components get developed separately and in isolation. At work something like this I could assign to another developer, who would implement it during a sprint and return it back for integration.
The only way for a developer at this point to provide a level of confidence that the implementation meets the requirements is to write unit tests.

There are a number of different approaches and techniques for writing code, e.g. TDD expects a developer to start with tests first, and write implementation second.
I tend to write a 'happy path' implementation first, so I could get a feel for the flow, selected abstractions and API usability. Then I start with tests to close the remaining gaps (i.e. null checks, edge cases etc).

Of course none of the above negates the usefulness of integration tests, but with integration tests it is impractical write tests for 'unhappy paths' or test edge cases. 

I hope this all makes sense.


> 0f98e02, 7ed9553 - the failed tests are expected, the implementation has changed. No suprises here.

Why there is no surprise for me? 
Each implementation is an abstraction. It may be a high level abstraction (e.g. _'table' is a piece of furniture_), a low level (e.g. _table consists of four wooden legs, table top, four brackets, eight bolts_) or anything in between. As well as each implementation is coded against certain requirements and assumptions. If these abstraction, assumptions or requirements change, then the component must be updated. Naturally if implementation changes, so do unit tests.
If component is implemented in line with OO design principles (i.e. SOLID) changes to the component should not affect its consumers. If they do - the component needs to be refactored further. 

> Why I would have to correct them? How can I know that I did not introduce any regression having to change 6 out of 8 tests?

As per above - if the component is SOLID, then changes to its implementation have no bearing on its consumers.
If you make change to a component then you have a reason to do so - either requirements or assumptions changed. And as a developer performing the change you are expected to have understanding of previous and target states.
Each unit test is supposed to test a particular code path. If a change affects several code paths, then it is expected that there would a number of tests affected. 
Setting commonly used mocks can/should be done in [SetUp], so changes to functionality and associated mocks can be done in one place.

**Comment by RussKie** [comment]
@jbialobr I have added an integration test, is there anything else blocking?

**Comment by RussKie** [comment]
Please have a look at 2b36088

**Comment by RussKie** [comment]
> I don't think that expanding relative path inside this controller goes along SRP. It should be implemented somewhere outside as it is a good candidate for being reused.

I'm currently refactoring the code to extract path full path resolution.
On Windows platform we are constrained to 248 character path limit (e.g. `Path.Combine`).
Are we ok with that, or is there some clever hack we need to apply?

**Comment by jbialobr** [comment]
>Are we ok with that, or is there some clever hack we need to apply?

I am ok with that.

File: GitUI/CommandsDialogs/FormCommit.cs [hunk 9]
```diff
@@ -222,19 +231,27 @@ private void ConfigureMessageBox()
...
+            {
+                Message.Text = _commitTemplate = _commitTemplateManager.LoadGitCommitTemplate();
+                return;
+            }
+            catch (FileNotFoundException ex)
+            {
+                MessageBox.Show(this, string.Format(_templateNotFound.Text, ex.FileName),
+                    _templateNotFoundCaption.Text, MessageBoxButtons.OK, MessageBoxIcon.Error);
+            }
+            catch (Exception ex)
```
Comment (RussKie) [inline GitUI/CommandsDialogs/FormCommit.cs]: should we handle `PathTooLongException` separately?

Comment (jbialobr) [reply GitUI/CommandsDialogs/FormCommit.cs]: It depends on how helpful you want to be. There can be also an exception related to invalid chars. 
If you want to be super helpful then I think `LoadGitCommitTemplate` should wrap its body in try-catch block and throw a new exception providing actual `commit.template` value and the caught exception's message. Then we could have a single try catch block here. But I am also fine with the existing solution.

Comment (RussKie) [reply GitUI/CommandsDialogs/FormCommit.cs]: I leave it as is now, in necessary we can revisit it later.
The general exception handler will relay the error message to the user.
